### PR TITLE
CEP 43: Conditional dependencies

### PR DESCRIPTION
**Pull request by jaimergp**
## Checklist for submitter

- [X] I am submitting a new CEP: [Conditional dependencies](https://github.com/conda/ceps/blob/conditional-deps/cep-XXXX.md).
  - [X] I am using the CEP template by creating a copy `cep-0000.md` named `cep-XXXX.md` in the root level.
- [ ] I am submitting modifications to CEP XX. 
  - [ ] The PR title reflects the CEP I'm modifying: `CEP XX: Amend XYZ`.
  - [ ] I updated the "Updated" date.
  - [ ] I added the link of this PR to the Discussions row.
  - [ ] I added or extended the `## Changelog` section right above the final "Copyright" section with an item that uses syntax `YYYY-MM-DD: Brief explanation of changes`.
- [ ] Something else: (add your description here).

## Checklist for CEP approvals

- [x] The vote period has ended and the vote has passed the necessary quorum and approval thresholds.
- [x] A new CEP number has been minted. Usually, this is `${greatest-number-in-main} + 1`.
- [x] The `cep-XXXX.md` file has been renamed accordingly.
- [x] The `# CEP XXXX - ` header has been edited accordingly.
- [x] The CEP status in the table has been changed to approved.
- [x] The last modification date in the table has been updated accordingly.
- [x] The table in the README has been updated with the new CEP entry.
- [ ] The `pre-commit` checks are passing.

Comes from https://github.com/conda/ceps/pull/111

### CODE REVIEW

**Comment by jaimergp** [comment]
Dear @conda/steering-council,

The vote for this CEP has started. It will be open for two weeks, until **May 11th, 2026, 23:59 Anywhere on Earth**. This time period has been chosen to make it eligible for time-out rules. As an Enhancement Proposal vote, it requires 60% affirmative votes to pass.

To vote, please mark the relevant checkbox under your username:

- Uwe Korn (@xhochy)
  - [x] yes
  - [ ] no
  - [ ] abstain
- Marius van Niekerk (@mariusvniekerk)
  - [x] yes
  - [ ] no
  - [ ] abstain
- Cheng H. Lee (@chenghlee)
  - [x] yes
  - [ ] no
  - [ ] abstain
- Filipe Fernandes (@ocefpaf)
  - [x] yes
  - [ ] no
  - [ ] abstain
- Marcelo Duarte Trevisani (@marcelotrevisani)
  - [ ] yes
  - [ ] no
  - [x] abstain
- Michael Sarahan (@msarahan)
  - [x] yes
  - [ ] no
  - [ ] abstain
- Marcel Bargull (@mbargull)
  - [x] yes
  - [ ] no
  - [ ] abstain
- John Kirkham (@jakirkham)
  - [ ] yes
  - [ ] no
  - [ ] abstain
- Jannis Leidel (@jezdez)
  - [x] yes
  - [ ] no
  - [ ] abstain
- Wolf Vollprecht (@wolfv)
  - [x] yes
  - [ ] no
  - [ ] abstain
- Jaime Rodríguez-Guerra (@jaimergp)
  - [x] yes
  - [ ] no
  - [ ] abstain
- Bas Zalmstra (@baszalmstra)
  - [x] yes
  - [ ] no
  - [ ] abstain
- Hind Montassif (@Hind-M)
  - [x] yes
  - [ ] no
  - [ ] abstain
- Pavel Zwerschke (@pavelzw)
  - [x] yes
  - [ ] no
  - [ ] abstain

File: cep-0043.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+## Specification
+
+This CEP extends [CEP 29](./cep-0029.md) with a new keyword, `when`.
+
+### Syntax
+
+A conditional dependency is defined as a `MatchSpec` string that features a `when` keyword, the value of which MUST be a string that encodes a logical expression of one or more `MatchSpec` queries. The logical expression follows a Python-like syntax: `MatchSpec` strings MAY be joined with operators `and` (logical AND) and `or` (logical OR), and grouped within parentheses `()` for precedence overrides.
+
+The inner `MatchSpec` queries inside `when` MUST be expressed in their square brackets syntax, with the exception of simple `name` and `version` queries that MAY be expressed as `{name}{operator}{version}` (no space separators). These inner `MatchSpec` queries MUST NOT feature their own `when` field.
```
Comment (jezdez) [inline cep-0043.md:57]: @chenghlee your scope-restriction concern from https://github.com/conda/ceps/pull/111#discussion_r3095111749 didn't carry over when this CEP got split out. That thread is still unresolved (baszalmstra asked for concrete ambiguous cases, and jaimergp's follow-up on whether `{name}{operator}{version}` covers forms like `python>=3,<4` is open too). Since voting is happening this week, could you re-raise or close it out on this PR?

**Comment by jezdez** [comment]
@jaimergp I have concerns about this vote being called if the thread above wasn't closed, you even asked about it?

**Comment by jaimergp** [comment]
It's a minor clarification, I'd say.

File: cep-0043.md (hunk 1 above)
Comment (wolfv) [reply cep-0043.md:57]: We asked Cheng about examples and he didn't reply :( We think the parser works fine, and we have tests for a number of edge cases.

Comment (baszalmstra) [reply cep-0043.md:57]: @jezdez I'm sorry my comment read passive-aggressive, it was not meant that way! We have already written a parser for this that works by quoting and escaping values. We didn't run into any difficult cases. I was asking about edge cases @chenghlee had in mind because I was curious whether there are difficult cases we missed. Since we didn't receive any during the RFC period, we felt it was fine to call a vote regardless.

Comment (chenghlee) [reply cep-0043.md:57]: Sorry; got caught up in other things. I'm okay with my comment on the previous CEP being unresolved; it's mostly that strictly speaking, the current square bracket notation supports ridiculous (but semantically meaningless) things like `[and=or foo="(123|4)" and 1]`, but having thought about it, I think the proper fix is to clamp down on allowable expressions in square brackets than to block this CEP.

**Comment by wolfv** [comment]
Mid voting period reminder, @msarahan @marcelotrevisani @jakirkham @mariusvniekerk @chenghlee @pavelzw

**Review by mariusvniekerk — APPROVED** [review]

**Comment by wolfv** [comment]
@mbargull @jakirkham last reminder to vote!

**Comment by benmoss** [comment]
I think the examples in this CEP are confusing, we aren't expecting users installing packages to really be specifying conditional dependencies, are we? I think the target audience for this feature would be package **authors** if I understand correctly.

I guess the framing here is that the same MatchSpec is used by both user types, but I'm not sure if that's by necessity or just incidental.

**Comment by jaimergp** [comment]
# Vote results

🗳️ The votes are in! Here is the tally for this @conda/steering-council vote. Thanks to everyone who participated.

🎉 **Result: PASSED**

## Summary

| Metric | Value | Threshold | Status |
|---|---|---|---|
| Quorum (participation) | 13/14 (92.8%) | ≥ 9 voters | ✅ |
| Approval (yes / yes+no) | 12/12 (100.00%) | ≥ 60% | ✅ |

_Abstentions count toward quorum but are excluded from the approval fraction._

## Tally

- ✅ **Yes:** 12 (100.00% of decisive)
- ❌ **No:** 0 (0.00% of decisive)
- 🤷 **Abstain:** 1
- 💤 **Not voted:** 1
- ⚠️ **Invalid:** 0

**Comment by jaimergp** [comment]
@benmoss, yes and no. I expect way more users on the packaging side (authoring recipes) then on the CLI (typing `conda create ...`), but I do see a use case for `environment.yml` specs instead of, for example, having separate files per OS for a CI pipeline.
